src/main/java/jalf/relation/algebra/Select.java

```diff
@@ -17,44 +20,49 @@
     private RelationType type;
 
     public Select(Relation operand, Selection selection, RelationType type) {
-        this.operand = operand;
-        this.selection = selection;
-        this.type = type;
+	this.operand = operand;
+	this.selection = selection;
     }
 
     public Select(Relation operand, Selection selection) {
-        this.operand = operand;
-        this.selection = selection;
-        this.type = typeCheck();
+	this.operand = operand;
+	this.selection = selection;
+	this.type = typeCheck();
     }
 
     @Override
     public RelationType getType() {
-        return type;
+	return type;
     }
 
     @Override
     protected RelationType typeCheck() {
-        return selection.toRelationType();
+	return selection.toRelationType();
     }
 
     @Override
     public <R> R accept(Visitor<R> visitor) {
-        return visitor.visit(this);
+	return visitor.visit(this);
     }
 
     @Override
     public Relation getOperand() {
-        return operand;
+	return operand;
     }
 
     public Selection getSelection() {
-        return selection;
+	return selection;
     }
 
     @Override
     public List<Object> getArguments() {
-        return Arrays.asList(selection);
+	return Arrays.asList(selection);
+    }
+
+    @Override
+    protected Keys lazyComputeKey() {
+	return new Keys(Key.candidate(AttrList.attrs(this.selection.gefns()
+		.keySet())));
```

Comment:
Add a `getLargestKey(s)` in `AbstractRelation` that delegates to `getLargestKey(s)` on `RelationType` that delegates to `toKey(s)` on `Heading`